R/db-odbc-oracle.R

```diff
@@ -0,0 +1,235 @@
+  sql_variant(
+    scalar = sql_translator(.parent = base_odbc_scalar,
+                            # Data type conversions are mostly based on this article
+                            # https://docs.oracle.com/cd/B19306_01/server.102/b14200/sql_elements001.htm
+                            as.character  = sql_cast("VARCHAR(255)"),
+                            as.numeric = sql_cast("NUMBER"),
+                            as.double = sql_cast("NUMBER"),
+                            is.null = function(x){
+                              build_sql(
+                                "CASE WHEN", x ," IS NULL THEN 1 ELSE 0 END "
+                              )},
+                            is.na = function(x){
+                              build_sql(
+                                "CASE WHEN", x ," IS NULL THEN 1 ELSE 0 END "
+                              )}
+    ) ,
+    base_odbc_agg,
+    base_odbc_win
+  )
+}
+
+
+#' @export
+sql_subquery.Oracle <- function(con, from, name = dbplyr:::unique_name(), ...) {
+  # Query fails if a quoted table is used.  Using sql() to remove quotes from an
+  # already 'idented' argument.
+  if (is.ident(from)) {
+    if (is.null(name)) {
+      build_sql("(", sql(from), ") ", con = con)
+    } else {
+      build_sql("(", sql(from), ") ", ident(name), con = con)
+    }
+  } else {
+    if (is.null(name)) {
+      build_sql("(", from, ")", con = con)
+    } else {
+      build_sql("(", from, ") ", ident(name), con = con)
+    }
+  }
+}
+
+#' @export
+db_analyze.Oracle <- function(con, table, ...) {
+  sql <- dbplyr::build_sql(
+    # Using ANALYZE TABLE instead of ANALYZE
+    # https://docs.oracle.com/cd/B28359_01/server.111/b28310/general002.htm#ADMIN11526
+    "ANALYZE TABLE ",
+    dbplyr::ident(table)
+    , con = con)
+  DBI::dbExecute(con, sql)
+}
+
+#' @export
+db_drop_table.Oracle <- function(con, table, force = FALSE, ...) {
+  # Oracle does not support the IF EXISTS argument passed by force = TRUE
+  # this is why we are not using the db_drop_table.OdbcConnection
+  sql <- dbplyr::build_sql(
+    "DROP TABLE ", dbplyr::sql(table),
+    con = con
+  )
+  DBI::dbExecute(con, sql)
+}
+
+#' @export
+db_list_tables.Oracle <- function(con)  {
+  # This function allows copy_to() to work.  This variance may
+  # be removed when the odbc package is able to handle a straight
+  # dbListTables
+  table_names <- dbGetQuery(con, "select table_name from dba_tables")
+  table_names$upper <- toupper(table_names$TABLE_NAME)
+  table_names <- table_names[table_names$TABLE_NAME == table_names$upper , 1]
+  as.character(table_names)
+}
+
+#' @export
+db_has_table.Oracle <- function(con, table){
+  # This function allows copy_to() to work.  This variance may
+  # be removed when the odbc package is able to handle a straight
+  # dbListTables
+  stopifnot(length(table) == 1)
+  toupper(table) %in% dbListTables(con)
+}
+
+#' @export
+db_copy_to.Oracle <- function(con, table, values,
```

Comment:
We can, although I have been trying to keep vender specific workarounds to the minimum. But Oracle's driver doesn't seem to support parameterized queries (at least the way that nanodbc / odbc does them), hence the below approach.